Cpp_hands_on_programming/lib/pipe/Ipc_pipe.cc

```diff
@@ -35,55 +50,103 @@ void Ipc_pipe::send() {
 			throw(std::runtime_error("Ipc_pipe::send: readFile"));
 		}
 		// write to pipe
-		if ((bytes_write = write(fdp, buffer.data(), bytes_read)) == -1) {
-	    	throw(std::runtime_error("Ipc_pipe::send: write to pipe"));
-		}
+		t.startTimer();   
+		do {
+			errno = 0;
+			signal(SIGPIPE, SIG_IGN); // ignore SIGPIPE in case read end is closed 
+			bytes_write = write(fdp, buffer.data(), bytes_read);
+			if (bytes_write == -1) {
+				if (errno == EAGAIN) {
+					t.checkTimer("Pipe is full for too long. Errno: " + std::string(strerror(errno)));
+				} else {
+					throw(std::runtime_error("Ipc_pipe::send: write. Errno: " + std::string(strerror(errno))));
+				}
+			}
+		} while (errno == EAGAIN);
 		// compare read and written bytes 
 		if (bytes_read == bytes_write) {
 			total += bytes_read;
 		} else {
 			throw(std::runtime_error("Ipc_pipe::send: read write difference"));
 		}
-	}
-
-	close(fdp);
+	}  
+ 	close(fdp);
 	std::cout << "File delivered by pipe successfully, exiting the program.." << std::endl;
 }
 
 void Ipc_pipe::receive() {
     std::cout << "Starting pipeReceive.." << std::endl;
 	int fdp, bytes;
-	off_t fileSize(0), total(0);
+	off_t fileSize(0), total(0), sendFileSize(0);
```

Comment:
I have done some research abouty off_t: it is a long => signed 2^32 => doesn't work with files larger than 2,1GB.

Reply:
What is the range of size_t?

Reply:
size_t max = 18446744073709551615